pyanaconda/modules/storage/reset.py

```diff
@@ -76,6 +77,12 @@ def _reload_modules(self):
 
         if arch.is_s390():
             zfcp.startup()
+            # Import persistent config of any s390 devices (dasd, zfcp, znet) from
+            # initrd to retain user choices made with rd.dasd, rd.zfcp, rd.znet.
+            execWithRedirect("chzdev",
```

Comment:
I could image to have all these calls in the `blivet` or `libblockdev` library and use their abstraction, but let's ask @vojtechtrefny first. Anyway, it is a very low-level code for Anaconda.

Reply:
> I could image to have all these calls in the `blivet` or `libblockdev` library and use their abstraction, but let's ask @vojtechtrefny first. Anyway, it is a very low-level code for Anaconda.

FWIW, this is by no means limited to handling block devices, so I'm not sure blivet or libblockdev would be a good home. The import is for persistent configuration of s390 devices of _any_ type, in particular including network.
I just used the hook here because we already had zfcp hooked here and this was promising for me and worked well during function test. I think we could even perform the import before anaconda starts, it just needs to be after switch root of the anaconda initrd, so we can import into the (volatile) root-fs of stage2 where anaconda runs in.

Reply:
@steffen-maier what exactly does the import do (applying the data)? Is it safe (sane) to call it multiple times?
The storage rescan can happen quite a few times during interactive installation (and the iscsi/fcoe/zfcp.startup() would do the startup only during the first call) so I think we should try to find a better place. Also given the import regards also network devices.

Reply:
https://github.com/ibm-s390-linux/s390-tools/pull/158/commits/7c21e53376238412ebb0977b5fec2cab8d2606fe#diff-c2758979d9762cd0f733650085a42de8533819b87c19665dc42f0c44e6442561 ("zdev/dracut: retain early persistent config over switch root") exports the persistent s390 device configuration established during early initrd (parsing result of rd.dasd, rd.zfcp, rd.znet) to `/run/zdev.initrd.config`, which is in a tmpfs that remains visible with its content even after dracut's switch-root. It's similar to and a generalization of anaconda-copy-s390ccwconf.sh (which this PR removes https://github.com/rhinstaller/anaconda/pull/5250/files/61aff0c3cc1fbedd142d869a9f8f106e723efdee#diff-f4a4c6c6e09f6db50c1840d9e254e2719ba33adae3878e7c9a1a99ed58e21d5c).

Here when we import the persistent config, we are already in the anaconda stage2 root-fs. Before the (first) import it has no s390 device persistent configuration. The import brings the persistent config into the stage2 environment.
An import **merges** with any persistent config that might already exist in the (stage2) environment. Hence, it should be safe to import multiple times (even if unnecessary).

Later during stage2, the user/kickstart can establish additional persistent config. Because other anaconda code (such as ZFCPDiskDevice.dracut_setup_args()) also needs to read any imported or established persistent config, the import needs to happen early enough.

Finally, the set union of persistent config gets persisted into the installation target (https://github.com/rhinstaller/anaconda/pull/5250/commits/022465897dac2d485c54cb019162642277df3d23#diff-d1f055451308b33c2271dfe2958a633fadc7a53acb9c5e7019ad1c45fc3ff818R293 ("write persistent config of any (dasd,zfcp,znet) s390 devices to sysrroot (#1802482,#1937049)")).) That kind of replaces the former different `...write...` methods for the different s390 device types.

I need help regarding a better code location for the import as that's way beyond my understanding of anaconda.

Reply:
Thank you for explanation, I was afraid the import does some activation or setup beyond the persistent config update. This way it seems safe and fine at this point. I will try to think about some better place.
And thank you for updating the tests!

Reply:
> Thank you for explanation, I was afraid the import does some activation or setup beyond the persistent config update. This way it seems safe and fine at this point. I will try to think about some better place. And thank you for updating the tests!

I wonder if it would make sense to import the state from initramfs by a systemd service (the services we use now: https://github.com/rhinstaller/anaconda/tree/master/data/systemd)

Reply:
https://github.com/rhinstaller/anaconda/pull/5250/commits/e94552e287c74cde37e330674ca93c38bd942d9b uses a systemd service and worked well

Reply:
Nice, thank you!